File: hammer-interaction/src/main/java/dev/vini2003/hammer/interaction/api/common/interaction/InteractionRule.java
package dev.vini2003.hammer.interaction.api.common.interaction;

import net.minecraft.tag.TagKey;

public class InteractionRule<T> {
	private final InteractionType type;
	private final InteractionMode mode;
	
	private final TagKey<T> tag;
	
	public InteractionRule(InteractionType type, InteractionMode mode, TagKey<T> tag) {
		this.type = type;
		this.mode = mode;
		
		this.tag = tag;
	}
	
	public InteractionType getType() {
		return type;
	}
	
	public InteractionMode getMode() {
		return mode;
	}
	
	public TagKey<T> getTag() {
		return tag;
	}
}

File: hammer-interaction/src/main/kotlin/dev/vini2003/hammer/interaction/common/interaction/InteractionRule.kt
package dev.vini2003.hammer.interaction.common.interaction

import dev.vini2003.hammer.common.util.serializer.TagKeySerializer
import dev.vini2003.hammer.common.util.serializer.UnsupportedSerializer
import kotlinx.serialization.Serializable
import net.minecraft.tag.TagKey

@Serializable
data class InteractionRule<out T>(
	val type: InteractionType,
	val mode: InteractionMode,
	val tagKey: @Serializable(with = TagKeySerializer::class) TagKey<@Serializable(with = UnsupportedSerializer::class) Any>
) {
	// Ignore is used to differentiate the two constructors, due to having the same JVM signature after type erasure.
	constructor(type: InteractionType, mode: InteractionMode, tagKey: TagKey<T>, ignore: Byte = 0) : this(type, mode, tagKey as TagKey<Any>)
}